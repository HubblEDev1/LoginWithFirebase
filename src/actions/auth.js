import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from "@firebase/auth";
import Swal from "sweetalert2";
import { googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

/*export const login = (uid, displayName) => {
    return {
        type: types.login
    }
}*/
export const startLoginEmailPasswords = (email, password) => {
    return (dispatch) => {
        /*setTimeout(() => {
            dispatch(login(123, 'edwin'))
        }, 3500);*/
        dispatch(startLoading());
        const auth = getAuth();
        signInWithEmailAndPassword(auth, email, password).then( ({user}) => {
            console.log(user);
            dispatch(finishLoading());
            dispatch(login(user.uid, user.displayName));
        }).catch((e) => {
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        })
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider).then(({user}) => {
            dispatch(login(user.uid, user.displayName));
        });
    }
}

export const startRegisterWithEmailPasswordName = (email, pass, name) => {
    return (dispatch) => {
        const auth = getAuth();
        createUserWithEmailAndPassword(auth, email, pass).then( async ({user}) => {
            await updateProfile(auth.currentUser, {displayName: name});
            dispatch(
                login(user.uid, user.displayName)
            );
        }).catch( e => {
            console.log(e);
            Swal.fire('Error', e.message, 'error');
        })
    }
}


export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid, 
        displayName
    }
}); 

export const startLogout = () => {
    return async ( dispatch ) => {
        const auth = getAuth();
        signOut(auth);
        dispatch( logout() );
    }
}

export const logout = () => ({
    type: types.logout,
})