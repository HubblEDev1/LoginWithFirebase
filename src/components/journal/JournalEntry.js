import React from 'react'

export const JournalEntry = () => {
    return (
        <div className="journal__entry">
            <div
                className="journal__entry-picture"
                style={{ 
                    backgroundSize: 'cover',
                    backgroundImage: 'url(https://i.pinimg.com/originals/98/40/4c/98404c70b9177d19259bbcbc2056fa61.jpg)'
                }}
            >
                
            </div>
            <div
                className="journal__entry-body"
            >
                <p
                    className="journal__entry-title"
                >
                    New day
                </p>
                <p className="journal__entry-content">
                    lorem ipsum dolor sit am
                </p>

            </div>
            <div className="journal__entry-date-box">
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}
