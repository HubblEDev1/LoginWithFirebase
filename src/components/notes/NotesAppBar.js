import React from 'react'

export const NotesAppBar = () => {
    return (
        <div className="notes__appbar">
            <span>20 August 1998</span>
            <div>
                <button className="btn">
                    Picture
                </button>
                <button className="btn">
                    Save
                </button>
            </div>
        </div>
    )
}
