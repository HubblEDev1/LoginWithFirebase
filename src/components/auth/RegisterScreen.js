import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';

import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    //const state = useSelector( (state) => state);
    const { msgError } = useSelector( state => state.ui );

    const [formValues, handleInputChange, reset] = useForm({
        name: '',
        email: '',
        pass: '',
        pass2: ''
    });

    const {name, email, pass, pass2} = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if( isFormValid() ){
            dispatch(startRegisterWithEmailPasswordName(email, pass, name));
        }
    }

    const isFormValid = () => {

        if(name.trim().length === 0){
            dispatch(setError('Name is required'));
            return false;
        }else if(!validator.isEmail(email)){
            dispatch(setError('Email is not valid'))
            return false;
        }else if( pass !== pass2 || pass < 5 ){
            dispatch(setError('Incorrect password'));
            return false;
        }

        dispatch(removeError());
        
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form onSubmit={handleRegister}>
                {
                    msgError && (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Name" 
                    name="name"
                    className="auth__input" 
                    value={name}    
                    onChange={handleInputChange}           
                />
                <input 
                    type="text"
                    placeholder="Email" 
                    name="email"
                    className="auth__input"
                    value={email} 
                    onChange={handleInputChange}                 
                />
                <input 
                    type="password"
                    placeholder="Password" 
                    name="pass"
                    className="auth__input"
                    value={pass} 
                    onChange={handleInputChange} 
                />
                <input 
                    type="password"
                    placeholder="Confirm password" 
                    name="pass2"
                    className="auth__input"
                    value={pass2} 
                    onChange={handleInputChange} 
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5" 
                >
                    Register
                </button>


                <Link
                    to="/auth/login"
                    className="link"
                >
                    Already Registered?
                </Link>
            </form>
        </>
    )
}
