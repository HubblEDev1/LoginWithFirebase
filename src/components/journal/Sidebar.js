import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    const dispatch = useDispatch();
    const {name} = useSelector( state => state.auth);

    console.log(name);

    const handleLogout = () => {
        dispatch( startLogout() );
    }


    return (
        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <h3>
                    <i className="fa fa-moon" />
                    <span>{name}</span>
                </h3>
                <button
                    className="btn"
                    onClick={handleLogout}
                >
                    Logout
                </button>
            </div>
            <div className="journal__new-entry">
                <i className="fa fa-calendar-plus fa-5x"></i>
                <p className="mt-5">New entry</p>
            </div>
            <JournalEntries/>
        </aside>
    )
}
