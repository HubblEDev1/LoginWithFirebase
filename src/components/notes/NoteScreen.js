import React from 'react'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
    return (
        <div
            className="notes__main-content"
        >
            <NotesAppBar />
            <div className="notes__content">
                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                />

                <textarea
                    placeholder="What happend today"
                    className="notes__textarea"
                >
                </textarea>

                <div className="notes__image">
                    <img
                        src="https://i.pinimg.com/originals/d8/cb/3c/d8cb3ce047d28819c65c4c5105fa30b3.jpg"
                        alt="Image"
                    />
                </div>

            </div>
        </div>
    )
}
