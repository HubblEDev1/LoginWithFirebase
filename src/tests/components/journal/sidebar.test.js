import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import '../../../setupTests';

import { mount } from "enzyme";
import { Provider } from "react-redux";

import { MemoryRouter } from 'react-router';
import { Sidebar } from '../../../components/journal/Sidebar';
import { startLogout } from '../../../actions/auth';
import { act } from '@testing-library/react';
import { getAuth, signInWithEmailAndPassword } from '@firebase/auth';
import { startNewNote } from '../../../actions/notes';

jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn(),
}));

jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {
        uid: '1', 
        name: 'Ed'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        active: {
            id: null,
        },
        notes: []
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store = { store }>
        <Sidebar/>
    </Provider>
);

describe('Testing sidebar', () => {

    beforeEach( () => {
        store = mockStore(initState);
        jest.clearAllMocks();
    } );
 
    test('should show correctly', () => {
        expect(wrapper).toMatchSnapshot();        
    });

    test('should call startLogout', async () => {
        wrapper.find('button').prop('onClick')();
        expect( startLogout ).toHaveBeenCalled();
    });

    test('should call startNewNote', async() => {
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect( startNewNote ).toHaveBeenCalled();
    });
    
})
