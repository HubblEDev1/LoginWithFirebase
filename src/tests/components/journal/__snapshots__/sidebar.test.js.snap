// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing sidebar should show correctly 1`] = `
<aside
  className="journal__sidebar animate__animated animate__fadeIn"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3>
      <i
        className="fa fa-moon"
      />
      <span>
        Ed
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="fa fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries animate__animated animate__fadeIn"
  />
</aside>
`;
